### ISSUE TITLE
Sign of pi_loss?

### ISSUE THREAD

**Issue by hholst80** [muupan/async-rl#22]
You are computing entropy in `policy_output.py` like:

```
- probs * log_probs
```

with a minus sign. This is expected to be positive (non-negative to be precise).

You are then computing `pi_loss` in `a3c.py` with a loop and subtracting terms:

```
for ...:
    pi_loss -= log_prob * advantage # sign (rhs) = sign(-advantage).
    pi_loss -= self.beta * entropy # sign (rhs) = 1. 
    v_loss += (v - R) ** 2 / 2
```

And finally you take `loss` as a (weighted) sum of `pi_loss` and `v_loss`.

Are you sure about this? It seems to me like you should add up `pi_loss` with `+=` on both the terms in the loop?

**Comment by hholst80** [comment]
On the other hand I think the purpose of the entropy in the `pi_loss` is to encourage high entropy actions. Do you agree? If so, we should minimize the negative entropy like you are doing. 

![image](https://cloud.githubusercontent.com/assets/6200749/17439440/6bc50e32-5b28-11e6-9859-096e5d4dc640.png)

NOTE: The maximum entropy is reached for `prob=exp(-1)` where `entropy==exp(-1)`.

**Comment by muupan** [comment]
I'm sure about that.
- `pi_loss` is what we want to minimize in optimization
- We want to maximize `A(s,a) log pi(a|s)`, so `-` is needed
- We want to maximize the entropy of `pi(a|s)`, so `-` is needed

**Comment by hholst80** [comment]
Thank you for your time and help to remove my confusion.

**Comment by hholst80** [comment]
While digging around with the signs we found a potential bug in the entropy calculation which can in some corner cases cause problems with NaN. (At least it did so for us...) The entropy function we're using now has a special case for p=0:

```python
def entropy(p, name=None):
  """Return H := sum(-p[i]*log(p[i])) for 0 <= p[i] <= 1."""
  p = tf.convert_to_tensor(p)
  zeros = tf.zeros_like(p)
  return tf.reduce_sum(
    tf.select(tf.equal(p, zeros), zeros, -p*tf.log(p)),
    reduction_indices=-1,
    name=name,
  )
```
